interop/ksp/src/main/kotlin/com/squareup/kotlinpoet/ksp/TypeParameterResolver.kt

```diff
@@ -0,0 +1,85 @@
+package com.squareup.kotlinpoet.ksp
+
+import com.google.devtools.ksp.symbol.KSType
+import com.google.devtools.ksp.symbol.KSTypeParameter
+import com.squareup.kotlinpoet.TypeVariableName
+
+/**
+ * A resolver for enclosing declarations' type parameters. Parent declarations can be anything with
+ * generics that child nodes declare as defined by [KSType.arguments].
+ *
+ * This is important for resolving inherited generics on child declarations, as KSP interop
+ * otherwise can't resolve them.
+ *
+ * In general, you want to retrieve an instance of this via [toTypeParameterResolver].
+ *
+ * @see toTypeParameterResolver
+ */
+@KotlinPoetKspPreview
+public interface TypeParameterResolver {
+  public val parametersMap: Map<String, TypeVariableName>
+  public operator fun get(index: String): TypeVariableName
+
+  public companion object {
+    /**
+     * An empty instance of [TypeParameterResolver], only should be used if enclosing declarations
+     * are known to not have arguments, such as top-level classes.
+     */
+    public val EMPTY: TypeParameterResolver = object : TypeParameterResolver {
+      override val parametersMap: Map<String, TypeVariableName>
+        get() = emptyMap()
+
+      override fun get(index: String): TypeVariableName {
+        error("No TypeParameter found for index $index")
```

Comment:
nit: should it throw something like `UnsupportedOperationException`?

Reply:
Not in this case I think, this is an unexpected condition rather than something unsupported or unimplemented

Reply:
Called out as based on the message it may seem that this API will return something for a different `index`, when in fact it will always throw.

Reply:
ahhh I didn't realize where you were commenting (was looking on mobile). Good point

Reply:
Went with NSEE to match what we do in metadata for this 74e323a